controller/share_manager_controller.go

```diff
@@ -1050,6 +1045,92 @@ func (c *ShareManagerController) getShareManagerTolerationsFromStorageClass(sc *
 	return tolerations
 }
 
+func (c *ShareManagerController) validateStorageNetworkForRWXVolume() (bool, error) {
+	storageNetwork, err := c.ds.GetSettingWithAutoFillingRO(types.SettingNameStorageNetwork)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to get setting value %v", types.SettingNameStorageNetwork)
+	}
+
+	storageNetworkForRWXVolumeEnabled, err := c.ds.GetSettingAsBool(types.SettingNameStorageNetworkForRWXVolumeEnabled)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to get setting value %v", types.SettingNameStorageNetworkForRWXVolumeEnabled)
+	}
+
+	return types.IsStorageNetworkForRWXVolume(storageNetwork, storageNetworkForRWXVolumeEnabled), nil
+}
+
+func (c *ShareManagerController) checkStorageNetworkApplied() (bool, error) {
+	targetSettings := []types.SettingName{types.SettingNameStorageNetwork, types.SettingNameStorageNetworkForRWXVolumeEnabled}
+	for _, item := range targetSettings {
+		if applied, err := c.ds.GetSettingApplied(item); err != nil || !applied {
+			return applied, err
+		}
+	}
+	return true, nil
+}
+
+func (c *ShareManagerController) canCleanupServiceAndEndpoint(ns, shareManagerName string) (bool, error) {
+	service, err := c.ds.GetService(c.namespace, shareManagerName)
+	if err != nil {
+		// if NotFound, means the service/endpoint is already cleaned up
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, errors.Wrap(err, "failed to get service")
+	}
```

Comment:
Is it possible the service is gone but the endpoint is still present?

Reply:
Honestly, the endpoint doesn't matter.
The main factor affecting the `status.endpoint` of ShareManager (w/o storage network) is the service.
The endpoint will be updated no matter if it is recreated or not.

But if we do not clean up the service file, the ClusterIP will already be set to `None`. This will block the next update for the ClusterIP (when we disable the storage network).

So the service is the main thing we need to take care of here.

cc @c3y1huang (I am afraid I got something misunderstanding)

Reply:
But you error out in `err = c.ds.DeleteKubernetesEndpoint(c.namespace, shareManager.Name)` if the deletion failed? Does it mean we don't need to error out and can just leave a warning message instead?

Reply:
Hmm, yeah, that's true.

I thought it would not be harmful if we deleted the endpoint with failure and did nothing next time.
But if we would like to keep the consistency, we could add extra checking for endpoint like a service.

WDYT?

Reply:
I discussed this with @c3y1huang and did some tests.

It looks like Kubernetes will clean up the `endpoint` once the corresponding `service` is removed.
So the case `service` deleted and 'endpoint` existing will be handled on the Kubernetes side.
It will not happen, I will also add the comment on the code.

Reply:
https://github.com/kubernetes/kubernetes/blob/v1.31.0/pkg/controller/endpoint/endpoints_controller.go#L374-L392

Reply:
Thanks for the code path reference. I updated it in the comment also!